zebra-scanner/src/tests.rs

```diff
@@ -0,0 +1,346 @@
+                .try_into()
+                .expect("position should fit in u32"),
+        };
+
+        let compact_block = block_to_compact(block, chain_metadata);
+
+        let res = scan_block(
+            &zcash_primitives::consensus::MainNetwork,
+            compact_block.clone(),
+            &vks[..],
+            &[(account, nf)],
+            None,
+        )
+        .unwrap();
+
+        transactions_found += res.transactions().len();
+        transactions_scanned += compact_block.vtx.len();
+        blocks_scanned += 1;
+
+        // scan backwards
+        if height.is_min() {
+            break;
+        }
+        height = height.previous()?;
+    }
+
+    // make sure all blocks and transactions were scanned
+    assert_eq!(blocks_scanned, 11);
+    assert_eq!(transactions_scanned, 11);
+
+    // no relevant transactions should be found
+    assert_eq!(transactions_found, 0);
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn scanning_from_fake_generated_blocks() -> Result<()> {
+    let account = AccountId::from(12);
+    let extsk = ExtendedSpendingKey::master(&[]);
+    let dfvk: DiversifiableFullViewingKey = extsk.to_diversifiable_full_viewing_key();
+    let vks: Vec<(&AccountId, &SaplingIvk)> = vec![];
+    let nf = Nullifier([7; 32]);
+
+    let cb = fake_compact_block(
+        1u32.into(),
+        BlockHash([0; 32]),
+        nf,
+        &dfvk,
+        1,
+        false,
+        Some(0),
+    );
+
+    // The fake block function will have our transaction and a random one.
+    assert_eq!(cb.vtx.len(), 2);
+
+    let res = scan_block(
+        &zcash_primitives::consensus::MainNetwork,
+        cb,
+        &vks[..],
+        &[(account, nf)],
+        None,
+    )
+    .unwrap();
+
+    // The response should have one transaction relevant to the key we provided.
+    assert_eq!(res.transactions().len(), 1);
+
+    Ok(())
+}
+
+// Convert a zebra block and meta data into a compact block.
+fn block_to_compact(block: Arc<Block>, chain_metadata: ChainMetadata) -> CompactBlock {
+    CompactBlock {
+        height: block
+            .coinbase_height()
+            .expect("verified block should have a valid height")
+            .0
+            .into(),
+        hash: block.hash().bytes_in_display_order().to_vec(),
+        prev_hash: block
+            .header
+            .previous_block_hash
+            .bytes_in_display_order()
+            .to_vec(),
+        time: block
+            .header
+            .time
+            .timestamp()
+            .try_into()
+            .expect("should work during 21st century"),
+        header: block
+            .header
+            .zcash_serialize_to_vec()
+            .expect("verified block should serialize"),
+        vtx: block
+            .transactions
+            .iter()
+            .cloned()
+            .enumerate()
+            .map(transaction_to_compact)
+            .collect(),
+        chain_metadata: Some(chain_metadata),
+
+        ..Default::default()
```

Comment:
Nitpick: Supplying default values for all new fields can cause errors or consensus bugs if the data structure changes later. Instead, supply specific placeholder values, so new fields cause compiler errors when the dependency is upgraded.

```suggestion
        // The protocol version is used for the gRPC wire format, so it isn't needed here.
        proto_version: 0,
```